flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -525,63 +525,99 @@ class ExprLowering {
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 4) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEsingle(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 10) {
+        llvm::APFloat floatVal{llvm::APFloatBase::x87DoubleExtended(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else if constexpr (KIND == 16) {
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEquad(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
+      } else {
+        // convert everything else to double
+        llvm::APFloat floatVal{llvm::APFloatBase::IEEEdouble(), str};
+        return genRealConstant<KIND>(builder.getContext(), floatVal);
       }
-      llvm_unreachable("real constant has no value");
     } else if constexpr (TC == Fortran::lower::ComplexCat) {
-      auto opt = con.GetScalarValue();
-      if (opt.has_value()) {
-        using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
-        return genval(Fortran::evaluate::ComplexConstructor<KIND>{
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->REAL()}},
-            Fortran::evaluate::Expr<TR>{
-                Fortran::evaluate::Constant<TR>{opt->AIMAG()}}});
-      }
-      llvm_unreachable("array of complex unhandled");
-    } else if constexpr (TC == Fortran::lower::CharacterCat) {
-      return genCharLit<KIND>(con.GetScalarValue().value(), con.LEN());
+      using TR = Fortran::evaluate::Type<Fortran::lower::RealCat, KIND>;
+      return genval(Fortran::evaluate::ComplexConstructor<KIND>{
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.REAL()}},
+          Fortran::evaluate::Expr<TR>{
+              Fortran::evaluate::Constant<TR>{value.AIMAG()}}});
     } else {
       llvm_unreachable("unhandled constant");
     }
   }
 
+  template <Fortran::common::TypeCategory TC, int KIND>
+  mlir::Value genArrayLit(
+      const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>>
+          &con) {
+    // TODO:
+    // - Multi-dimensional array
+    using T = Fortran::evaluate::Type<TC, KIND>;
+    if (con.Rank() > 1) {
+      TODO(); // "Multi dimensional array not yet supported"
+    }
+    // FIXME:
+    // Information about the array extents needs to be obtained from
+    // the symbol table
+    int64_t size = static_cast<int64_t>(con.size());
+    fir::SequenceType::Shape shape(1, static_cast<int64_t>(size));
+    auto arrayTy = fir::SequenceType::get(shape, converter.genType(TC, KIND));
+    auto idxTy = builder.getIndexType();
+    mlir::Value array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
+    // FIXME:
+    // Once we have the right array extents from symbols, the correct
+    // lower and upper bounds can be used here in place of 1 and size.
```

Comment:
Looking a bit more at the `Constant<T>` code, there is an easy way to iterate through `Constant<T>` that will work with arbitrary rank:
```
Fortran::evaluate::ConstantSubscripts subscripts = con.lbounds();
do {
  auto constant = genScalarLit<TC, KIND>(con.At(subscripts));
  // ... 
} while (con.IncrementSubscripts(subscripts));
```
You still need to do the plumbing between Fortran subscripts (lbound based) and subscripts used in `InsertValueOp` though.